File: src/prod/src/ServiceModel/management/FaultAnalysisService/Chaos/GetChaosReportDescription.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------


#include "stdafx.h"

using namespace Common;
using namespace Naming;
using namespace ServiceModel;
using namespace std;

using namespace Management::FaultAnalysisService;

StringLiteral const TraceComponent("GetChaosReportDescription");

GetChaosReportDescription::GetChaosReportDescription()
: filterSPtr_()
, continuationToken_()
, clientType_()
{
}

GetChaosReportDescription::GetChaosReportDescription(
std::shared_ptr<ChaosReportFilter> & filter,
std::wstring & token,
std::wstring & clientType)
: filterSPtr_(filter)
, continuationToken_(token)
, clientType_(clientType)
{
}

GetChaosReportDescription::GetChaosReportDescription(GetChaosReportDescription const & other)
: filterSPtr_(other.filterSPtr_)
, continuationToken_(other.continuationToken_)
, clientType_(other.clientType_)
{
}

GetChaosReportDescription::GetChaosReportDescription(GetChaosReportDescription && other)
: filterSPtr_(move(other.filterSPtr_))
, continuationToken_(move(other.continuationToken_))
, clientType_(move(other.clientType_))
{
}

// This never gets called when client is REST
Common::ErrorCode GetChaosReportDescription::FromPublicApi(
    FABRIC_GET_CHAOS_REPORT_DESCRIPTION const & publicDescription)
{
    ChaosReportFilter filter;
    auto error = filter.FromPublicApi(*publicDescription.Filter);

    if (!error.IsSuccess())
    {
        Trace.WriteError(TraceComponent, "filter.FromPublicApi failed, error: {0}.", error);
        return error;
    }

    filterSPtr_ = make_shared<ChaosReportFilter>(filter);

    HRESULT hr = StringUtility::LpcwstrToWstring(
        publicDescription.ContinuationToken,
        true, /*accept nullptr*/
        0, /*IntPtr.Zero from managed translates into "\0", a string of length 0*/
        ParameterValidator::MaxStringSize,
        continuationToken_);

    if (FAILED(hr))
    {
        Trace.WriteWarning(TraceComponent, "GetChaosReportDescription::FromPublicApi/Failed to parse ContinuationToken: {0}", hr);
        return ErrorCode::FromHResult(hr);
    }

    if (publicDescription.Reserved != nullptr) // managed
    {
        auto descriptionClientType = reinterpret_cast<FABRIC_CHAOS_CLIENT_TYPE*>(publicDescription.Reserved);

        if (descriptionClientType->ClientType != nullptr)
        {
            hr = StringUtility::LpcwstrToWstring(
                descriptionClientType->ClientType,
                true, /*accept nullptr*/
                0, /*IntPtr.Zero from managed translates into "\0", a string of length 0*/
                ParameterValidator::MaxStringSize,
                clientType_);

            if (FAILED(hr))
            {
                Trace.WriteWarning(TraceComponent, "GetChaosReportDescription::FromPublicApi/Failed to parse ClientType: {0}", hr);
                return ErrorCode::FromHResult(hr);
            }
        }
    }
    else // native
    {
        clientType_ = L"native";
    }

    return ErrorCodeValue::Success;
}

void GetChaosReportDescription::ToPublicApi(
    __in Common::ScopedHeap & heap,
    __out FABRIC_GET_CHAOS_REPORT_DESCRIPTION & result) const
{
    auto publicReportFilterPtr = heap.AddItem<FABRIC_CHAOS_REPORT_FILTER>();
    filterSPtr_->ToPublicApi(heap, *publicReportFilterPtr);
    result.Filter = publicReportFilterPtr.GetRawPointer();

    result.ContinuationToken = heap.AddString(continuationToken_);

    auto clientTypeEx1 = heap.AddItem<FABRIC_CHAOS_CLIENT_TYPE>();

    if (!clientType_.empty())
    {
        clientTypeEx1->ClientType = heap.AddString(clientType_);
    }
    else
    {
        clientTypeEx1->ClientType = nullptr;
    }

    result.Reserved = clientTypeEx1.GetRawPointer();
}

File: src/prod/src/managed/Api/src/System/Fabric/Description/GetChaosReportDescription.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Description
{
    using System;
    using System.Fabric.Chaos.Common;
    using System.Fabric.Chaos.DataStructures;
    using System.Fabric.Common;
    using System.Fabric.Interop;
    using System.Fabric.Strings;
    using System.Globalization;

    internal sealed class GetChaosReportDescription
    {
        public GetChaosReportDescription(
            ChaosReportFilter filter,
            string continuationToken)
        {
            this.Filter = filter;
            this.ContinuationToken = continuationToken;
        }

        public ChaosReportFilter Filter
        {
            get;
            private set;
        }

        public string ContinuationToken
        {
            get;
            internal set;
        }

        internal string ClientType { get; set; }

        public override string ToString()
        {
            return string.Format(CultureInfo.InvariantCulture, "Filter={0}, ContinuationToken: {1}, ClientType: {2}",
                this.Filter,
                this.ContinuationToken,
                this.ClientType);
        }

        internal IntPtr ToNative(PinCollection pinCollection)
        {
            var nativeGetChaosReportDescription = new NativeTypes.FABRIC_GET_CHAOS_REPORT_DESCRIPTION();

            nativeGetChaosReportDescription.Filter = this.Filter.ToNative(pinCollection);
            nativeGetChaosReportDescription.ContinuationToken = pinCollection.AddObject(this.ContinuationToken);
            var clientType = new NativeTypes.FABRIC_CHAOS_CLIENT_TYPE
            {
                ClientType = pinCollection.AddObject(ChaosConstants.ManagedClientTypeName)
            };

            nativeGetChaosReportDescription.Reserved = pinCollection.AddBlittable(clientType);

            return pinCollection.AddBlittable(nativeGetChaosReportDescription);
        }

        internal static unsafe GetChaosReportDescription CreateFromNative(IntPtr nativeRaw)
        {
            NativeTypes.FABRIC_GET_CHAOS_REPORT_DESCRIPTION native = *(NativeTypes.FABRIC_GET_CHAOS_REPORT_DESCRIPTION*)nativeRaw;

            var filter = ChaosReportFilter.FromNative(native.Filter);
            string continuationToken = NativeTypes.FromNativeString(native.ContinuationToken);

            var description = new GetChaosReportDescription(filter, continuationToken);

            if (native.Reserved != IntPtr.Zero)
            {
                var clientType = *((NativeTypes.FABRIC_CHAOS_CLIENT_TYPE*)native.Reserved);

                if (clientType.ClientType != IntPtr.Zero)
                {
                    description.ClientType = NativeTypes.FromNativeString(clientType.ClientType);
                }
            }

            return description;
        }
    }
}
